igvc_platform/src/imu/YostLabDriver.cpp

```diff
@@ -171,62 +215,67 @@ void YostLabDriver::run()
+          imu_msg_.orientation_covariance = { .1, 0, 0, 0, .1, 0, 0, 0, .1 };
 
-        // Filtered orientation estimate
-        tf::quaternionTFToMsg(quat, imu_msg_.orientation);
-        imu_msg_.orientation_covariance = { .1, 0, 0, 0, .1, 0, 0, 0, .1 };
+          // Corrected angular velocity.
+          Eigen::Vector3d angular_vel_raw(parsed_val_[4], parsed_val_[5], parsed_val_[6]);
+          angular_vel_raw = correction_mat * angular_vel_raw;
 
-        // Corrected angular velocity.
-        Eigen::Vector3d angular_vel_raw(parsed_val_[4], parsed_val_[5], parsed_val_[6]);
-        angular_vel_raw = correction_mat * angular_vel_raw;
+          // Corrected linear acceleration.
+          Eigen::Vector3d linear_accel_raw(parsed_val_[7], parsed_val_[8], parsed_val_[9]);
+          linear_accel_raw = correction_mat * linear_accel_raw * GRAVITY;
 
-        // Corrected linear acceleration.
-        Eigen::Vector3d linear_accel_raw(parsed_val_[7], parsed_val_[8], parsed_val_[9]);
-        linear_accel_raw = correction_mat * linear_accel_raw * GRAVITY;
+          imu_msg_.angular_velocity.x = angular_vel_raw[0];
+          imu_msg_.angular_velocity.y = angular_vel_raw[1];
+          imu_msg_.angular_velocity.z = angular_vel_raw[2];
+          imu_msg_.angular_velocity_covariance = { .1, 0, 0, 0, .1, 0, 0, 0, .07 };
 
-        imu_msg_.angular_velocity.x = angular_vel_raw[0];
-        imu_msg_.angular_velocity.y = angular_vel_raw[1];
-        imu_msg_.angular_velocity.z = angular_vel_raw[2];
-        imu_msg_.angular_velocity_covariance = { .1, 0, 0, 0, .1, 0, 0, 0, .07 };
+          imu_msg_.linear_acceleration.x = linear_accel_raw[0];
+          imu_msg_.linear_acceleration.y = linear_accel_raw[1];
+          imu_msg_.linear_acceleration.z = linear_accel_raw[2];
+          imu_msg_.linear_acceleration_covariance = { .1, 0, 0, 0, .1, 0, 0, 0, .1 };
 
-        imu_msg_.linear_acceleration.x = linear_accel_raw[0];
-        imu_msg_.linear_acceleration.y = linear_accel_raw[1];
-        imu_msg_.linear_acceleration.z = linear_accel_raw[2];
-        imu_msg_.linear_acceleration_covariance = { .1, 0, 0, 0, .1, 0, 0, 0, .1 };
-
-        double sensor_temp = parsed_val_[10];
+          sensor_temp_ = parsed_val_[10];
 
+          this->imu_pub_.publish(imu_msg_);
+          this->lastUpdateTime_ = ros::Time::now();
+          this->last_quat_ = quat;
+        }
+        else
+        {
+          ROS_WARN_STREAM("Incomplete message from IMU. Throwing it away.");
+        }
         parsed_val_.clear();
-        this->imu_pub_.publish(imu_msg_);
-
-        double roll, pitch, yaw;
-        tf::Matrix3x3(quat).getRPY(roll, pitch, yaw);
-
-        ROS_INFO_THROTTLE(1.0, "[YostLabImuDriver] R: %f, P: %f, Y: %f -- IMU Temp: %f F ", roll, pitch, yaw,
-                          sensor_temp);
+        line_num_ = 0;
       }
     }
     loop_rate.sleep();
     ros::spinOnce();
+    this->updater.update();
   }
```

Comment:
This is a really long function. I think it'd be much nicer if you refactored it into like multiple functions. Maybe something (vaguely) like

```
auto parsed_vals = parseMessage();
if (!parsed_vals.complete) {
  ROS_WARN_STREAM(...)
}
```

or something.